static/js/message_list_view.js

```diff
@@ -132,6 +132,85 @@ function update_message_date_divider(opts) {
         timerender.render_date(curr_time, prev_time, today)[0].outerHTML;
 }
 
+function clear_unread_messages_divider(msg) {
+    msg.want_unread_divider = false;
+    msg.unread_divider_html = undefined;
+}
+
+function add_unread_messages_divider(msg) {
+    msg.want_unread_divider = true;
+    msg.include_sender = true;
+
+    var NEW_MESSAGES_TEXT = 'NEW MESSAGES';
+    var unread_divider_html = $('<span></span>');
+
+    if (msg.want_date_divider) {
+        var date_divider_html = $(msg.date_divider_html);
+        var timerender_class = date_divider_html.attr('class');
+        unread_divider_html.attr('class', timerender_class);
+        unread_divider_html.append(date_divider_html.html());
+        unread_divider_html.find('.date-line').after(
+            NEW_MESSAGES_TEXT, '<hr class="date-line">');
+    } else {
+        unread_divider_html.text(NEW_MESSAGES_TEXT);
+    }
+
+    msg.unread_divider_html = unread_divider_html[0].outerHTML;
+}
+
+function update_unread_messages_divider(opts) {
+    var prev_msg_container = opts.prev_msg_container;
+    var curr_msg_container = opts.curr_msg_container;
+    var filter = opts.list.data.filter;
+
+    // Our unread message dividers, as currently conceived, don't really make sense in
+    // interleaved narrows, since we could have read messages after unread messages in
+    // those contexts.
+    // As a result, we currently don't offer these "New messages" notices in those views.
+    var is_topic = filter.has_operator('stream') && filter.has_operator('topic');
+    var is_pm = filter.has_operator('pm-with') || filter.has_operator('group-pm-with');
+
+    if (!(is_topic || is_pm)) {
+        clear_unread_messages_divider(curr_msg_container);
+        return;
+    }
+
+    if (!prev_msg_container.msg.unread && curr_msg_container.msg.unread) {
+        add_unread_messages_divider(curr_msg_container);
+        return;
+    }
+
+    clear_unread_messages_divider(curr_msg_container);
+}
+
+function clear_message_divider(msg) {
+    clear_message_date_divider(msg);
+    clear_unread_messages_divider(msg);
+
+    msg.want_divider = false;
+    msg.divider_html = undefined;
+}
+
+function update_message_divider(opts) {
+    var prev_msg_container = opts.prev_msg_container;
+    var curr_msg_container = opts.curr_msg_container;
+
+    if (!prev_msg_container) {
+        clear_message_divider(curr_msg_container);
+        return;
+    }
+
+    update_message_date_divider(opts);
+    // Unread messages divider should be updated once the date divider has been updated
+    update_unread_messages_divider(opts);
+
+    curr_msg_container.want_divider = curr_msg_container.want_date_divider ||
+                                      curr_msg_container.want_unread_divider;
+    // Unread divider html is preferred over the date divider html
+    curr_msg_container.divider_html = curr_msg_container.unread_divider_html ||
+                                      curr_msg_container.date_divider_html;
```

Comment:
I think this structure is an improvement, but I think we can go further: I think it should be possible to merge the `update_message_date_divider` and `update_unread_messages_divider` functions (basically just compute once whether the dates and different and whether this is the unread boundary, and if so, just do a `templates.render` on a single shared HTML template that renders dividers, with conditionals to handle the differences between the two very similar divider styles).  Does that make sense?

Reply:
Yeah! I'll make this change.